File: tools/llm-project-mapper/src/walk.js
import { readdirSync, readFileSync, statSync } from 'node:fs';
import { extname, join } from 'node:path';

import { languageFor } from './languages.js';

/** @typedef {import('./types.js').FileEntry} FileEntry */
/** @typedef {import('./types.js').TreeNode} TreeNode */

const NULL_BYTE = 0;

/**
 * @param {Buffer} buf
 * @returns {boolean} true if the buffer looks like binary content
 */
function looksBinary(buf) {
  const limit = Math.min(buf.length, 8000);
  for (let i = 0; i < limit; i++) {
    if (buf[i] === NULL_BYTE) return true;
  }
  return false;
}

/**
 * @param {Buffer} buf
 * @returns {number}
 */
function countLines(buf) {
  if (buf.length === 0) return 0;
  let lines = 0;
  for (let i = 0; i < buf.length; i++) {
    if (buf[i] === 0x0a) lines++;
  }
  if (buf[buf.length - 1] !== 0x0a) lines++;
  return lines;
}

/**
 * Recursively walk a project tree, honouring an ignore predicate.
 *
 * @param {string} root absolute project root
 * @param {Object} [options]
 * @param {(relPath: string, name: string, isDir: boolean) => boolean} [options.ignore]
 * @param {number} [options.maxDepth] tree depth retained in the structure
 * @param {number} [options.maxFileBytes] skip line counting above this size
 * @returns {{tree: TreeNode, files: FileEntry[],
 *            metrics: {totalFiles: number, totalDirs: number,
 *                      totalLines: number, totalBytes: number,
 *                      maxDepth: number}}}
 */
export function walkProject(root, options = {}) {
  const ignore = options.ignore ?? (() => false);
  const maxDepth = options.maxDepth ?? 8;
  const maxFileBytes = options.maxFileBytes ?? 2_000_000;

  /** @type {FileEntry[]} */
  const files = [];
  const metrics = {
    totalFiles: 0,
    totalDirs: 0,
    totalLines: 0,
    totalBytes: 0,
    maxDepth: 0,
  };

  /**
   * @param {string} absDir
   * @param {string} relDir
   * @param {number} depth
   * @returns {TreeNode}
   */
  function visit(absDir, relDir, depth) {
    metrics.maxDepth = Math.max(metrics.maxDepth, depth);
    /** @type {TreeNode} */
    const node = {
      name: relDir === '' ? '.' : relDir.split('/').pop() || relDir,
      type: 'dir',
      children: [],
    };

    let names = [];
    try {
      names = readdirSync(absDir).sort((a, b) => a.localeCompare(b));
    } catch {
      return node;
    }

    for (const name of names) {
      const abs = join(absDir, name);
      const rel = relDir === '' ? name : relDir + '/' + name;
      let st;
      try {
        st = statSync(abs);
      } catch {
        continue;
      }
      const isDir = st.isDirectory();
      if (ignore(rel, name, isDir)) continue;

      if (isDir) {
        metrics.totalDirs++;
        const child = visit(abs, rel, depth + 1);
        if (depth + 1 > maxDepth) {
          node.children?.push({ name, type: 'dir', truncated: true });
        } else {
          node.children?.push(child);
        }
      } else if (st.isFile()) {
        const ext = extname(name).toLowerCase();
        const language = languageFor(name, ext);
        const bytes = st.size;
        /** @type {number|null} */
        let lines = null;
        if (bytes <= maxFileBytes) {
          try {
            const buf = readFileSync(abs);
            if (!looksBinary(buf)) lines = countLines(buf);
          } catch {
            lines = null;
          }
        }
        metrics.totalFiles++;
        metrics.totalBytes += bytes;
        if (lines != null) metrics.totalLines += lines;
        files.push({ path: rel, language, bytes, lines });
        if (depth + 1 <= maxDepth) {
          node.children?.push({ name, type: 'file', language, bytes, lines });
        }
      }
    }
    return node;
  }

  const tree = visit(root, '', 0);
  return { tree, files, metrics };
}

File: tools/llm-project-mapper/src/walk.rs
//! Directory traversal: builds the structure tree and the flat file list.
//! Line counting is performed separately (in parallel) by `count_lines`.

use std::fs;
use std::path::Path;

use crate::ignore::Ignore;
use crate::languages::language_for;
use crate::types::{FileEntry, TreeNode};

pub struct WalkResult {
    pub tree: TreeNode,
    pub files: Vec<FileEntry>,
    pub total_files: u64,
    pub total_dirs: u64,
    pub total_bytes: u64,
    pub max_depth: u64,
}

/// Walk `root` (cheap: only readdir + metadata), honouring `ignore`. Files keep
/// their byte size; `lines` is filled later by the parallel counting pass.
pub fn walk(root: &Path, ignore: &Ignore, max_depth: u64) -> WalkResult {
    let mut result = WalkResult {
        tree: TreeNode::dir("."),
        files: Vec::new(),
        total_files: 0,
        total_dirs: 0,
        total_bytes: 0,
        max_depth: 0,
    };
    let tree = visit(root, "", 0, ignore, max_depth, &mut result);
    result.tree = tree;
    result
}

fn visit(
    abs_dir: &Path,
    rel_dir: &str,
    depth: u64,
    ignore: &Ignore,
    max_depth: u64,
    out: &mut WalkResult,
) -> TreeNode {
    out.max_depth = out.max_depth.max(depth);
    let name = if rel_dir.is_empty() {
        ".".to_string()
    } else {
        rel_dir.rsplit('/').next().unwrap_or(rel_dir).to_string()
    };
    let mut node = TreeNode::dir(name);

    let mut names: Vec<String> = match fs::read_dir(abs_dir) {
        Ok(rd) => rd
            .filter_map(|e| e.ok())
            .map(|e| e.file_name().to_string_lossy().to_string())
            .collect(),
        Err(_) => return node,
    };
    names.sort();

    for name in names {
        let abs = abs_dir.join(&name);
        let rel = if rel_dir.is_empty() {
            name.clone()
        } else {
            format!("{}/{}", rel_dir, name)
        };
        let meta = match fs::symlink_metadata(&abs) {
            Ok(m) => m,
            Err(_) => continue,
        };
        let is_dir = meta.is_dir();
        if ignore.is_ignored(&rel, &name, is_dir) {
            continue;
        }

        if is_dir {
            out.total_dirs += 1;
            // Always recurse so metrics and the flat file list (with stable,
            // sequential file ids) stay complete; only the retained tree is
            // gated by depth.
            let child = visit(&abs, &rel, depth + 1, ignore, max_depth, out);
            if depth >= max_depth {
                let mut truncated = TreeNode::dir(name);
                truncated.truncated = true;
                node.children.push(truncated);
            } else {
                node.children.push(child);
            }
        } else if meta.is_file() {
            let ext = extension(&name);
            let language = language_for(&name, &ext);
            let bytes = meta.len();
            let file_id = out.files.len();
            out.total_files += 1;
            out.total_bytes += bytes;
            out.files.push(FileEntry {
                path: rel.clone(),
                abs_path: abs.clone(),
                language: language.to_string(),
                bytes,
                lines: None,
                file_id,
            });
            if depth < max_depth {
                node.children.push(TreeNode {
                    name,
                    is_file: true,
                    language: Some(language.to_string()),
                    bytes: Some(bytes),
                    lines: None,
                    children: Vec::new(),
                    truncated: false,
                    file_id: Some(file_id),
                });
            }
        }
    }
    node
}

fn extension(name: &str) -> String {
    match name.rfind('.') {
        Some(i) if i > 0 => name[i..].to_lowercase(),
        _ => String::new(),
    }
}

/// Count newline-terminated lines in a file, or `None` if it is binary or
/// cannot be read. Files larger than `max_bytes` are skipped by the caller.
pub fn count_lines(path: &Path) -> Option<i64> {
    let data = fs::read(path).ok()?;
    let limit = data.len().min(8000);
    if data[..limit].contains(&0) {
        return None; // binary
    }
    if data.is_empty() {
        return Some(0);
    }
    let mut lines = data.iter().filter(|&&b| b == b'\n').count() as i64;
    if *data.last().unwrap() != b'\n' {
        lines += 1;
    }
    Some(lines)
}
